Enable shortenIDs to recycle existing IDs

nthykier opened the pull request:
This patch enables shortenIDs to remap IDs currently in use.  This is
very helpful to ensure that scour does not change been "optimal" and
"suboptimal" choices for IDs as observed in GH#186.

Closes: #186
Signed-off-by: Niels Thykier <niels@thykier.net>

oberstet commented:
@nthykier oh, cool: I just noticed that you have a number of PRs queuing .. thanks for contributing! @Ede123 how is your current time budget? =) if none, I see what I can do @nthykier to review it

nthykier commented:
Tobias Oberstein:
> @nthykier oh, cool: I just noticed that you have a number of PRs queuing .. thanks for contributing!

No problem.  I am glad that some of these changes are useful :)

Thanks,
~Niels

Ede123 commented on scour/scour.py:
```diff
@@ -822,34 +823,44 @@ def renameID(doc, idFrom, idTo, identifiedElements, referencedIDs):
     return num
 
 
+def _protectedIDs(seenIDs, options):
```
You should name this `protected_ids` for consistency with the other function (which I feel is more important than to mark this internal, for which the long-term goal is something along the lines of #121).

Also note we try to use PEP8 naming for new code where reasonable (don't force it though - only if it makes sense and doesn't introduce any inconsistencies). Naming similar things similarly is always more important than following any guidelines.

nthykier commented:
Eduard Braun:
> Ede123 requested changes on this pull request.
> 
> General idea and most of the code looks fine, however it causes a serious regression:

Hi,

Thanks for the review. :)

I have pushed fixes as separate commits on top of the branch.  Let me
know if (and in that case, when) you want the branch rebased/squashed.

> Since we can now re-use IDs already present in the document we need to make sure to properly rename them in the process (I'm not sure what the best possibility for this is, especially in a way that does not potentially regress performance - i.e. consider a case where you add one heavily referenced object to an already scoured document which would require Scour to rename all IDs - so this part needs some investigation).
> 

Added in d6406a34706e7ae1422595a5eda6d8bc7b50b1a8.

To be honest, I think the major win here is making scour produce
"stable" output once it has picked shortened IDs.  The performance
optimization is only applicable in a "best case" scenario where optimal
IDs have already been handed out.

Technically, I have solved it by ensuring that scour does not change the
ID of a node unless the ID changes size.

> Testcase:
> ```
> [...]
> ```
> 

Thanks for that.  I have added a unit test using that to avoid regressions.

>> @@ -822,34 +823,44 @@ def renameID(doc, idFrom, idTo, identifiedElements, referencedIDs):
>                      node.setAttribute(attr, newValue)
>                      num += len(oldValue) - len(newValue)
>  
> -        del referencedIDs[idFrom]
> -        referencedIDs[idTo] = referringNodes
> +        # We deliberately leave referencedIDs alone to enable us to bulk update
> +        # IDs where two nodes swap IDs.  (GH#186)
> +        # del referencedIDs[idFrom]
> +        # referencedIDs[idTo] = referringNodes
> 
> I don't understand this comment, can you clarify?
> 

I have deleted it and have replaced it with
91503c6d7e0882ebfff7bb52def7938a63e9864e

>>  
>      return num
>  
>  
> +def _protectedIDs(seenIDs, options):
> 
> You should name this `protected_ids` for consistency with the other function (which I feel is more important than to mark this internal, for which the long-term goal is something along the lines of #121).
> 

Ok, done in 00cf42b554e11e09348218b9e66d72de178694ae

> Also note we try to use PEP8 naming for new code where reasonable (don't force it though - only if it makes sense and doesn't introduce any inconsistencies). Naming similar things similarly is always more important than following any guidelines.
> 

This seems like a tricky balance that I am sure I will get wrong again
eventually. :)

Thanks,
~Niels

Ede123 commented:
>  Technically, I have solved it by ensuring that scour does not change the ID of a node unless the ID changes size.

Sounds reasonable. It does not really matter wether we assign "a" or "z" as an ID from a technical point of view.  
The only downside I can think of is that scouring the original file (with a small modification) vs. scouring a previously scoured file (with the same small modification) would result in different IDs, but I assume that's hardly relevant in practice.

> Thanks for that. I have added a unit test using that to avoid regressions.

I'm afraid I did not describe the regression well enough (and my testcase wasn't designed well enough either). The problem occurs when actually re-writing is in order, see this updated minimal testcase, which still reveals the behavior I wanted to show before:

```SVG
<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
 <defs>
  <rect id="a" width="80" height="50" fill="red"/>
  <rect id="long_id" width="80" height="50" fill="blue"/>
 </defs>
 <use xlink:href="#a"/>
 <use xlink:href="#long_id"/>
 <use xlink:href="#long_id"/>
</svg>
```

The "long_id" gets rewritten to "a"!

Yet another testcase which might be close to "real world" files which were scoured and later modified: [testcase.svg.txt](https://github.com/scour-project/scour/files/1913258/testcase.svg.txt)  
In this case the new algorithm also rewrites "long_id" "a" and even manages to update the previous link to "a" with "aa", but fails to update the actual IDs, so it effectively exchanges the objects.

> This seems like a tricky balance that I am sure I will get wrong again eventually. :) 

Well, in general PEP8 names are the way we want to go, so if in doubt, use them. I'm only hesitant to rewrite most of the code just to follow PEP8 names (changes like these make it unnecessarily hard to browse the git history using "git blame" while not really improving anything).

nthykier commented:
Eduard Braun:
>>  Technically, I have solved it by ensuring that scour does not change the ID of a node unless the ID changes size.
> 
> Sounds reasonable. It does not really matter wether we assign "a" or "z" as an ID from a technical point of view.  

Indeed.  That is what I thought.

By the way, I noticed that scour's ID allocator only uses a-z, but
according the the SVG standard (via the XML standard) allows for a lot
more characters in the ID field.

https://www.w3.org/TR/2008/REC-xml-20081126/#NT-Name says:

> [4]   	NameStartChar	   ::=   	":" | [A-Z] | "_" | [a-z] | [#xC0-#xD6] | [#xD8-#xF6] | [#xF8-#x2FF] | [#x370-#x37D] | [#x37F-#x1FFF] | [#x200C-#x200D] | [#x2070-#x218F] | [#x2C00-#x2FEF] | [#x3001-#xD7FF] | [#xF900-#xFDCF] | [#xFDF0-#xFFFD] | [#x10000-#xEFFFF]
> [4a]   	NameChar	   ::=   	NameStartChar | "-" | "." | [0-9] | #xB7 | [#x0300-#x036F] | [#x203F-#x2040]
> [5]   	Name		   ::=   	NameStartChar (NameChar)*

(from https://www.w3.org/TR/2008/REC-xml-20081126/#id, which is found
via https://www.w3.org/TR/SVG/struct.html#Core.attrib)

In other words, we can at least use "_", uppercase letters.
Furthermore, if we handle the first character differently from the rest,
then we can use 0-9, "-" and "." as well (without having to look at all
the UTF-8 characters also permitted).


> The only downside I can think of is that scouring the original file (with a small modification) vs. scouring a previously scoured file (with the same small modification) would result in different IDs, but I assume that's hardly relevant in practice.
> 

In the general case, we cannot optimally shorten IDs and still produce
identical output if people alter the reference counts of the image
between runs of scour.

Obviously, we can still avoid "pointless" reassignments or the
"avalanche-effect" when we assign an existing ID a shorter name.

>> Thanks for that. I have added a unit test using that to avoid regressions.
> 
> I'm afraid I did not describe the regression well enough (and my testcase wasn't designed well enough either). The problem occurs when actually re-writing is in order, see this updated minimal testcase, which still reveals the behavior I wanted to show before:
> 
> ```SVG
> <?xml version="1.0" encoding="UTF-8"?>
> <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
>  <defs>
>   <rect id="a" width="80" height="50" fill="red"/>
>   <rect id="long_id" width="80" height="50" fill="blue"/>
>  </defs>
>  <use xlink:href="#a"/>
>  <use xlink:href="#long_id"/>
>  <use xlink:href="#long_id"/>
> </svg>
> ```
> 
> The "long_id" gets rewritten to "a"!
> 

Thanks for the extended test case. :)

> Yet another testcase which might be close to "real world" files which were scoured and later modified: [testcase.svg.txt](https://github.com/scour-project/scour/files/1913258/testcase.svg.txt)  
> In this case the new algorithm also rewrites "long_id" "a" and even manages to update the previous link to "a" with "aa", but fails to update the actual IDs, so it effectively exchanges the objects.
> 

Ok, will have a look at fixing this issue.

Thanks,
~Niels

nthykier commented:
@Ede123 I have written 039022e, which I believe addresses your concerns.

nthykier commented:
Hey @Ede123

Let me know if there is anything I can do to improve the changes in this pull request.

JoKalliauer commented:
Thanks to @nthykier : I can confirm it is fixed.
